[WIP] Convert EasyNetQ to .NET Core [feature/netcore]

micdenny opened the pull request:
closes #508 connected to https://github.com/EasyNetQ/EasyNetQ.Management.Client/issues/34

@zidad @Pliner this follows the GitFlow style.

Now we can push on the branch `feature/netcore` to also update this PR, then when we're happy with the feature, we merge on `develop`.

Then we could configure appveyor to push:

`master` to **nuget.org** as _stable_
`develop` to **nuget.org** as _pre-release_
`feature` branch to **myget** as _pre-release_

useful resources:
- https://docs.microsoft.com/en-us/dotnet/articles/standard/library
- https://docs.microsoft.com/en-us/dotnet/articles/core/tutorials/libraries
- https://github.com/dotnet/corefx/blob/master/Documentation/architecture/net-core-applications.md
- https://blogs.msdn.microsoft.com/cesardelatorre/2016/06/28/running-net-core-apps-on-multiple-frameworks-and-what-the-target-framework-monikers-tfms-are-about/

zidad commented:
@micdenny excellent thanks, looks good to me, shall I merge?

micdenny commented:
I would work on feature branch instead of merge on develop, actually is the same but a cleaner approach, now the feature branch is on official repo so we can work pushing on it without other PR and more important this PR remains active until we finish the development.

I have some problem compiling the core project, should I add other nuget repo than the public nuget?

micdenny commented:
I rebase this on top of `develop`, save your work on another branch, and `pull --f` before continue working on.

zidad commented:
Perfect, thanks!

micdenny commented:
> I have some problem compiling the core project, should I add other nuget repo than the public nuget?

now I can compile, was some mistake on my local repo config

of course the dependent project DI.* need a local disk repo at the moment to compile against a dev package, but I will try to manually push a package on myget, to let it build w/o other steps

micdenny commented:
I don't know why I can't add successfully a project reference from DI.Tests project to Di.Autofac that is now a netcore, but this cross reference should be supported, in fact you can add the project reference and visual studio shows the correct link versus the correct 4.5.1 assembly, but does not work, it gives you a compile error, but if you add the reference using the browse it works like expected, so to now I leaved the browse reference instead project.

Then I pushed the first ENQ netcore package on myget and can be used.

micdenny commented:
https://docs.microsoft.com/en-us/dotnet/articles/core/tutorials/using-on-windows

in section `A mixed .NET Core library and .NET Framework application` there's why you cannot add the reference as project reference, simply it is not possible, also there they suggest you to `browse` for the dll or add as nuget package!!

micdenny commented:
@zidad It is mandatory to compile against `net451` instead `net45`?

micdenny commented:
@zidad ok sorry, I realized now that `RabbitMQ.Client` is compiled against `net451` so yes, it is mandatory :smile:

zidad commented:
Yep! Oh and I did mention we couldn't reference the new assemblies in old projects, that was part of the reason I proposed to switch to xunit, that works well with multiple targeting + .NET core

micdenny commented:
now we need to understand if we want to support net45, that means we have to stick the net45 to the old `RabbitMQ.Client` and compile a separated nuget package

or

starting from v2, just compile against net451 and rabbitmq.client >=4

zidad commented:
I prefer not to keep supporting the older versions. People can always install older versions of ENQ, but could we really think of reason why someone wouldn't be able to update from 4.5 to 4.5.1? It has been included in Windows Update for ages :)

micdenny commented:
I agree with you, even if with big projects retargetting could be not so easy, for instance I have a product with more than 300 projects to upgrade, but is something I have to do, so is good that ENQ sticks on newer framework, even if I have to admit that supporting also net45 could be not so difficult, probably a really low effort when our porting netcore will be finished and compiled by appveyor, adding one more target could be trivial.

Let's see after we complete the migration.

micdenny commented on Source/EasyNetQ.Core/project.json:
```diff
@@ -0,0 +1,136 @@
+      "imports": "dnxcore50",
+      "buildOptions": {
+        "define": [
+          "NET_CORE"
```
we can avoid this some symbols are already generated:

https://docs.microsoft.com/en-us/dotnet/articles/core/tutorials/libraries
```
The build system is aware of the following preprocessor symbols used in #if directives:+

.NET Framework 2.0   --> NET20
.NET Framework 3.5   --> NET35
.NET Framework 4.0   --> NET40
.NET Framework 4.5   --> NET45
.NET Framework 4.5.1 --> NET451
.NET Framework 4.5.2 --> NET452
.NET Framework 4.6   --> NET46
.NET Framework 4.6.1 --> NET461
.NET Framework 4.6.2 --> NET462
.NET Standard 1.0    --> NETSTANDARD1_0
.NET Standard 1.1    --> NETSTANDARD1_1
.NET Standard 1.2    --> NETSTANDARD1_2
.NET Standard 1.3    --> NETSTANDARD1_3
.NET Standard 1.4    --> NETSTANDARD1_4
.NET Standard 1.5    --> NETSTANDARD1_5
.NET Standard 1.6    --> NETSTANDARD1_6
```

zidad commented:
The symbols idea was here from last year (before it was Core) already from an earlier PR. All these symbols weren't documented/available yet.

But I must say I don't fully understand what the differences between all the different targets are, but I don't see a preprocessor symbol in that list that covers what we have now, basically 1 symbol for everything  before .NET Framework 4.6.3 and everything after .NET Framework 4.6.3. 

I'd like to prevent having multiple symbols in each #if statement (core/standard/etc), if you have a proposal that can make that happen I'm all for it! Otherwise I don't mind having a simple abstraction with our own symbols for this.

micdenny commented:
> basically 1 symbol for everything before .NET Framework 4.6.3 and everything after .NET Framework 4.6.3 (core/standard).

you're right, I will investigate if they export something more, otherwise we leave ours

micdenny commented:
https://docs.microsoft.com/en-us/dotnet/articles/core/tutorials/libraries

> If you want to target the .NET Framework versions 4.0 or below, or you wish to use an API available in the .NET Framework but not in the .NET Standard (for example, System.Drawing), read the following sections and learn how to multitarget.

If we want to support `>=net451` as far as I understood is sufficient to compile against the framework `netstandard1.2` leaving the dependencies on `NETStandard.Library` to `1.6.0` because is compatible.

In fact I can use the dll produced in the netstandard1.2 folder on a net451 project, actually also in a net45 without complaining, but I think because the public interface is the same `NETStandard.Library` to `1.6.0` but then the implementation will differ and some methods can throw an exception at runtime.

micdenny commented:
@zidad what are all the empty `#if !NET_CORE` in the code?

```csharp
#if !NET_CORE
#endif
```

micdenny commented:
> If we want to support >=net451 as far as I understood is sufficient to compile against the framework netstandard1.2 leaving the dependencies on NETStandard.Library to 1.6.0 because is compatible.

nope, `rabbitmq.client` supports only:

- .NETFramework 4.5.1
- .NETStandard 1.5

micdenny commented:
@zidad I'm also working on a personal branch if you want to get a look at it:

https://github.com/micdenny/EasyNetQ/tree/feature/micdenny-netcore


I've removed build of `netcoreapp1.0`, clean the empty `#if` and removed the old `EasyNetQ NET45` and shared project, leaving only the net core project with all the source files in there, plus I've convertered one more DI project in net core.

I did all the changes in separated commit, so if we don't like some of those, I can rebase and drop it.

micdenny commented:
I've finished `EasyNetQ.DI.*` unfortunately `Castle.Windsor` does not support `netstandard` so I converted the project in netcore, but it compiles only `net451`.

Sunday I will push everything on this branch as well.

micdenny commented:
pushed some work

zidad commented:
Wow that's a lot of great work!

sysmatrix1 commented:
Thanks for your help. We have been using this library extensively in our .NET core project since rc1. After converting our projects to support .NET core 1.0 - This is one of the very few libraries which don't have support. Any ideas when it will be available through public nuget feed?

takinosaji commented:
Excuse me. When are you going to release latest RabbitMQ.Client support?

zidad commented:
@takinosaji if you like you can try the latest prerelease package. You can udpate to it with this command:

`update-package easynetq -pre -source https://www.myget.org/F/easynetq-unstable/api/v2 -version 2.0.3-netcore0001`

BrettJaner commented:
The above package feed url didn't work for me.  The pre-release packages seemed to have moved here: https://www.myget.org/F/easynetq-unstable/api/v3/index.json

zidad commented:
I've tried to do some final fixes on the beta version for a release to nuget, however I can't seem to build any 1.0 .NET core project anymore on my machine due to the installation of .net core 1.1. I'm not really willing to fix it all for the gazillionth time, nor updating it to .NET Core 1.1, which requires all project files to be rewritten (again). 

I'm really sorry for everyone that has put effort in converting this to .NET core, but considering this is all spare time, I'm not willing to waste a minute more on any .NET Core project until the formats stabilize and has proper support from at least CLI tools, build servers and versioning strategies for the new formats. 

I've just bled out from being on the way-too-painful bleeding edge of .NET these days :)
